rbd-mirror:  Update the sync point object number during sync

rjfd opened the pull request:
Fixes: http://tracker.ceph.com/issues/15108

dillaman commented on src/tools/rbd_mirror/image_sync/ImageCopyRequest.cc:
```diff
@@ -198,6 +200,11 @@ void ImageCopyRequest<I>::handle_object_copy(int r) {
       }
     }
 
+    if (!m_updating_sync_point &&
```
I'd rather see the SafeTimer be utilized to periodically update the sync point every X seconds (new config option).

dillaman commented on src/tools/rbd_mirror/image_sync/ImageCopyRequest.cc:
```diff
@@ -210,6 +217,48 @@ void ImageCopyRequest<I>::handle_object_copy(int r) {
+    return;
+  }
+
+  update_progress("UPDATE_SYNC_POINT");
```
No need to update the progress since this is an out-of-band state.

dillaman commented on src/tools/rbd_mirror/image_sync/ImageCopyRequest.cc:
```diff
@@ -210,6 +217,48 @@ void ImageCopyRequest<I>::handle_object_copy(int r) {
+  }
+
+  Mutex::Locker l(m_lock);
+  m_updating_sync_point = false;
```
Need to handle the cases where the state machine completes / cancels while there is an in-flight sync point update. In such a case, you cannot complete the state machine until the sync point callback has been invoked / processed.

dillaman commented on src/test/rbd_mirror/image_sync/test_mock_ImageCopyRequest.cc:
```diff
@@ -396,6 +438,67 @@ TEST_F(TestMockImageSyncImageCopyRequest, Cancel) {
+
+  ASSERT_TRUE(complete_object_copy_sync(mock_object_copy_request, 0, 0, &ctx_0));
+  ASSERT_TRUE(complete_object_copy_sync(mock_object_copy_request, 1, 0, &ctx_1));
+  ASSERT_TRUE(complete_object_copy_sync(mock_object_copy_request, 2, 0, &ctx_2));
```
There is a simpler way to inject the cancel request exactly when object 2 is copied -- modify MockObjectCopyRequest to add a dummy ```construct``` mock method that it's ```create``` method invokes.  Then you can EXPECT_CALL on that ```construct(<object number>)``` and invoke the cancel method directly from the expectation.

rjfd commented:
@dillaman addressed all your comments and pushed the new changes.

Regarding the unit test problem, I could not do the way you suggested as when the `create` method is called, we are holding the `m_lock` from `ImageCopyRequest` instance, and that prevent us from calling any other method that tries to acquire the same lock again.

dillaman commented:
In the Invoke, you should be able to enqueue a lambda on the op_work_queue to inject cancel and complete the provide Context.

rjfd commented:
In the last push, I changed the implementation of the test cases, and I think it's more simple now.

dillaman commented on src/tools/rbd_mirror/image_sync/ImageCopyRequest.h:
```diff
@@ -100,6 +100,11 @@ class ImageCopyRequest : public BaseRequest {
 
+  bool m_updating_sync_point;
+  Context *m_update_sync_ctx;
+  // TODO: make interval configurable
```
Minor: TODO

rjfd replied:
Should I use the `md_config_obs_t` and create a new config option for this?

dillaman commented on src/tools/rbd_mirror/image_sync/ImageCopyRequest.cc:
```diff
@@ -205,6 +220,88 @@ void ImageCopyRequest<I>::handle_object_copy(int r) {
   if (complete) {
+    bool do_flush = true;
+    {
+      Mutex::Locker l(*m_timer_lock);
```
Minor: l -> timer_locker

dillaman commented on src/tools/rbd_mirror/image_sync/ImageCopyRequest.cc:
```diff
@@ -146,7 +148,20 @@ void ImageCopyRequest<I>::send_object_copies() {
+
+  if (m_update_sync_ctx) {
+    Mutex::Locker l(*m_timer_lock);
+    m_timer->add_event_after(m_update_sync_point_interval,
```
You might have to schedule this while holding ```m_timer_lock``` and ```m_lock``` to avoid a potential race if it completes when you drop the lock.

rjfd replied:
I don't think I need to hold both locks, although I need to acquire `m_timer_lock` before testing `m_update_sync_ctx != nullptr`.
If the sync completes, and the code starting at line 227 starts to execute,  before line 158, then the code starting in line 227 (which is holding `m_timer_lock`), will assign `m_update_sync_ctx = nullptr`. Then after line 158 we acquire the `m_timer_lock` but we will not call `add_event_after` anymore.

dillaman replied:
Well, if ```complete``` was false on line 150 and when you drop the lock ```handle_object_copy``` invokes and discovers it should complete. At this point, ```m_update_sync_ctx``` is not null, ```handle_object_copy``` cancels the event (which hasn't been scheduled yet), sets it to null (future revision), and continues.  Therefore, you would need to grab the timer lock before testing m_update_sync_ctx for null  on like 159.

rjfd replied:
Yep, that's what I said in my comment above, and already fixed in the implementation, although still didn't pushed the changes

dillaman commented on src/tools/rbd_mirror/image_sync/ImageCopyRequest.cc:
```diff
@@ -205,6 +220,88 @@ void ImageCopyRequest<I>::handle_object_copy(int r) {
+      Mutex::Locker locker(m_lock);
+      if (!m_updating_sync_point) {
+        if (m_update_sync_ctx != nullptr) {
+          m_timer->cancel_event(m_update_sync_ctx);
```
m_update_sync_ctx = nullptr;

dillaman commented on src/tools/rbd_mirror/image_sync/ImageCopyRequest.cc:
```diff
@@ -205,6 +220,88 @@ void ImageCopyRequest<I>::handle_object_copy(int r) {
+  }
+
+  if (m_sync_point->object_number
+      && (m_object_no-1) == m_sync_point->object_number.get()) {
```
Minor: move && to previous line

dillaman commented on src/tools/rbd_mirror/image_sync/ImageCopyRequest.cc:
```diff
@@ -205,6 +220,88 @@ void ImageCopyRequest<I>::handle_object_copy(int r) {
+
+  if (schedule_update) {
+    Mutex::Locker l(*m_timer_lock);
+    m_timer->add_event_after(m_update_sync_point_interval, m_update_sync_ctx);
```
I think this would be the same race condition if it completed as soon as you dropped the lock.

rjfd replied:
In this case I have to acquire `m_timer_lock` and then test if `m_update_sync_ctx != nullptr`, because it could already have been assign to `nullptr` in line 232 (after pushing the new changes).

rjfd commented:
@dillaman done pushing new changes

dillaman commented on src/common/config_opts.h:
```diff
@@ -1235,6 +1235,11 @@ OPTION(rbd_journal_object_flush_bytes, OPT_INT, 0) // maximum number of pending
+/**
+ * RBD Mirror options
+ */
+OPTION(rbd_mirror_sync_point_update_interval, OPT_DOUBLE, 1) // number of seconds between each update of the image sync point object number
```
Minor: can we switch "interval" to "age" and bump the default to something like 30 seconds?

dillaman commented on src/tools/rbd_mirror/image_sync/ImageCopyRequest.cc:
```diff
@@ -146,7 +149,22 @@ void ImageCopyRequest<I>::send_object_copies() {
+
+  {
+    Mutex::Locker timer_locker(*m_timer_lock);
+    if (m_update_sync_ctx) {
```
I think this is going to cause helgrind warning since you sometimes hold only ```m_timer_lock``` when touching ```m_update_sync_ctx```, sometimes you hold ```m_lock```, and sometimes you hold both.

rjfd commented:
@dillaman pushed new changes

dillaman commented:
@rjfd Did you change anything?

rjfd commented:
@dillaman yes, I changed the name and default value of config option. I also ran the unit test against helgrind but didn't got any warning.

rjfd commented:
@dillaman rebased on top of master
